### ISSUE TITLE
React and JSX

### ISSUE THREAD

**Issue by Leninsky** [electron-userland/electron-webpack#60]
```
 "devDependencies": {
    "babel-preset-react": "^6.24.1",
    "electron": "1.7.9",
    "electron-builder": "^19.42.1",
    "electron-webpack": "1.10.1",
    "webpack": "^3.8.1"
  }
}
```
in package.json

I am using 
```javascript
import React from 'react'
import ReactDOM from 'react-dom'
import App from './app.jsx'

ReactDOM.render(<App />, document.getElementById('app'))
```bash
[./src/renderer/index.js] ./src/renderer/index.js 171 bytes {0} [built]
      + 9 hidden modules

  ERROR in ./src/renderer/app.jsx
  Module parse failed: Unexpected token (7:16)
  You may need an appropriate loader to handle this file type.
  | export default class App extends Component {
  |      render() {
  |          return <h1> Hello </h1>
  |      }
  |
   @ ./src/renderer/index.js 3:0-28
   @ multi (webpack)-dev-server/client?http://localhost:64380 webpack/hot/dev-server ./src/renderer/index.js
  Child html-webpack-plugin for "index.html":
       1 asset
      [./node_modules/html-loader/index.js?minimize=false!../../../__virtual__/renderer-index.html] ./node_modules/html-loader?minimize=false!/__virtual__/renderer-index.html 367 bytes {0} [built]

┗ ----------------------------
┏ Renderer -------------------

  webpack: Failed to compile.

┗ ----------------------------
```

**Comment by loopmode** [comment]
Weird. Works for me without the custom config.

**Comment by loopmode** [comment]
Aaaah sorry, didn't see you use jsx extension. Works out of the box if you use .js instead

**Comment by walleXD** [comment]
You can also use `Babel-plugin-module-resolver` to add support for more extensions

**Comment by thealexbaron** [comment]
Spent a bunch of time trying to figure this out as well.

**Comment by jcity** [comment]
Yeah, I spend a while figuring this out as well. Here's an example using a babel plugin:

```js
module.exports = {
  module: {
    rules: [
      {
        test: /\.jsx?$/,
        loader: "babel-loader",
        options: {
          plugins: ['transform-decorators-legacy'],
          presets: ["es2015", "react", "stage-1"]
        }
      }
    ]
  },
  resolve: {
    extensions: ['.js', '.jsx']
  }
}
```

**Comment by walleXD** [comment]
@jcity If you use `babel-module-resolver` you can add extension  support using only `.babelrc` file

**Comment by brandonweiss** [comment]
This bit me as well. I'm very new to electron and fairly new to webpack, but not particularly new to react.

Is the file extension intentionally supposed to be `.js` for some reason? It does work, but if it has JSX in it then not using the `.jsx` extension throws my editor off.

**Comment by jcity** [comment]
Yeah, I ran into the same thing... If you're using JSX/TSX the file extension needs to be .jsx/.tsx respectively.

**Comment by loopmode** [comment]
@brandonweiss regardless of this specific issue - always try to let your editor know who's boss and who's not. :) 
I mean.. you shouldn't have to make project decisions to please an IDE or editor, it should be the other way round.

**Comment by deadcoder0904** [comment]
@loopmode Please add this to the docs as it was really confusing for me, but eventually I figured it out through Github issues

Beginners won't be able to find this so I think it should be mentioned in the **Add-ons>React JSX** docs

**Comment by starkos** [comment]
In case someone runs into this who, like me, also wants to use Typescript, it appears that you can sidestep this issue by using Typescript instead of Babel to process the .jsx/.tsx files. Don't install `@babel/react`. Instead, follow the instructions for [the Typescript add-on](
https://webpack.electron.build/add-ons#typescript), then [add JSX support](https://webpack.electron.build/add-ons#adding-typescript-support-for-jsx-tsx-files) to `tsconfig.json`.

*So far* everything (HMR, etc.) seems to be working correctly…not sure if this setup will cause issues down the road?

**Comment by cainvommars** [comment]
@deadcoder0904 Thanks for providing your solution. I ran into the same issue and your repo and comment here were very helpful.

**Comment by stereokai** [comment]
@starkos it is not working for me. Can't get JSX to work at all, not with Babel and not with Typescript.

**Comment by stereokai** [comment]
@starkos Typescript + TSX didn't work for me even though I followed the docs precisely.

@deadcoder0904 	this solution gets JSX to work but breaks HMR for me.

**Comment by naturalethic** [comment]
For me the issue was having `@babel/preset-react` in `dependencies`.  It must be placed in `devDependencies`.

**Comment by loopmode** [comment]
Any devDependencies - not regular dependencies - starting with @babel/preset or @babel/plugin (same with older babel-* variants) get automatically added to the babel-loader config in webpack.
